### ISSUE TITLE
CargoDetails() returning no data for mission commodities

### ISSUE THREAD

**Issue by Darkcyde13** [EDCD/EDDI#2280]
### Expected
CargoDetails() to provide appropriate data for mission related cargo, at all times.

### Observed
CargoDetails() returns no data for 'source and return' mission cargo after travelling to another system and relogging.

### Steps to reproduce
1. Take a 'source and return' mission.
2. Observe that CargoDetails() provides all appropriate data for the mission cargo.
3. Travel to another system.
4. Land at a station and log out of game and out of EDDI.
5. Run EDDI and load back into game.
6. Observe that CargoDetails() no longer returns any data for the mission cargo.

### Configuration
* Version [Please give full version numbers here, not just "latest" or "current"]
  * EDDI Version: V4.0.1
  * VoiceAttack Version: V1.8.9
* Localization [if applicable]
  * EDDI's language: British English
  * Voice: Ivona 2 Amy

### Investigation Notes
I've been updating my missions scripts, and running several while permit hunting.  Normally, I take and complete all my missions in one session, so as not to run out of time and fail any.  However, last night, I had to stop before completing all my missions, leaving a couple of data courier, a hack megaship, and a source & return mission.

On loading my game today, I discovered that EDDI had 'forgotten' all about the required cargo for the source & return mission.  I completed the missions I had and set about looking into why EDDI had no data for that mission.

I took another S&R mission, and ran the following script:
```
{for mission in missions:

  Mission: {mission.localisedname}.
  
  {set cargo to CargoDetails(mission.commodity)}
  name: {cargo.name}.
  category: {cargo.category}.
  owned: {cargo.owned}, {dump(cargo.owned)}, {type(cargo.owned)}.
  stolen: {cargo.stolen}, {dump(cargo.stolen)}, {type(cargo.stolen)}.
  haulage: {cargo.haulage}, {dump(cargo.haulage)}, {type(cargo.haulage)}.
  total: {cargo.total}, {dump(cargo.total)}, {type(cargo.total)}.
  need: {cargo.need}, {dump(cargo.need)}, {type(cargo.need)}.
  price: {cargo.price}, {dump(cargo.price)}, {type(cargo.price)}.

}
```
This is the output I got:
> Mission: Source and return 54 units of Pesticides. name: Pesticides. category: Chemicals. owned: 0, 0, number. stolen: 0, 0, number. haulage: 0, 0, number. total: 0, 0, number. need: 54, 54, number. price: 0, 0, number.

I then quit and reloaded EDDI and ran the script again, with the same output.  I then quit and reloaded Elite, ran the script, and again got the same expected output.  Finally, I quit both, rebooted my PC (leaving it off for 30 seconds), loaded EDDI and Elite, and ran the script again.  I still got the same expected output.  So I was rather confused and thought maybe I had just had a glitch.

However, I thought I'd try jumping to another system, as that's what I had done yesterday.  I went to a nearby system, landed at a station, quit both EDDI and Elite (back to desktop), and then reloaded both.  This then resulted in the problem occurring, giving me the following output:
> Mission: Source and return 54 units of Pesticides. name:. category:. owned:, , void. stolen:, , void. haulage:, , void. total:, , void. need:, , void. price:, , void.

So, it seems that somehow, after jumping to another system and relogging, EDDI is forgetting the data associated with mission cargo.

I went and purchased a few Pesticides to see what would happen, and got this result:
> Mission: Source and return 54 units of Pesticides. name: Pesticides. category: Chemicals. owned: 16, 16, number. stolen: 0, 0, number. haulage: 0, 0, number. total: 16, 16, number. need: 0, 0, number. price: 233, 233, number.

As you can see, the 'need' variable data is showing as zero, when it should be 54.

One final test, I sold them back to the station and ran the script again:

> Mission: Source and return 54 units of Pesticides. name:. category:. owned:, , void. stolen:, , void. haulage:, , void. total:, , void. need:, , void. price:, , void.

So now it's back to not knowing anything about the mission cargo.

Let me know if there is anything else you would like me to try.

### EDDI Logs
Logs of my play sessions today, although I'm not sure how much use they'll be: [EDDI Logs.zip](https://github.com/EDCD/EDDI/files/7429506/EDDI.Logs.zip)

### Player journals
Logs of my play sessions today: [ED Journals.zip](https://github.com/EDCD/EDDI/files/7429500/ED.Journals.zip)

**Comment by Tkael** [comment]
Huh. I'm a little baffled why jumping to a different system might do that but I'll see if I can puzzle it out. 

I think the data that you're looking for should be preserved in cargomonitor.json. If you could step through the process you described and determine at which step the information saved to cargomonitor.json is cleared (or if it not cleared) then that would be helpful information.

**Comment by Darkcyde13** [comment]
OK, so I've gone to test this out, and got some weirdness going on....

First of all, my cargomonitor,json file hasn't been changed since 21st November 2020!  So I renamed it, so that EDDI could create a new one.... but it didn't.  So, no cargomonitor.json file.  This would explain a few things! :)  It looks like EDDI hasn't written/updated a cargomonitor.json file for almost a year!  Now, it's entirely possible I hadn't updated EDDI when v3.7.1 came out, as I did have extended periods of time not playing.  So it may be since 3.7.1 came out that the file hasn't been written.  All the other EDDI files are written and updated in that folder, so it's not a permissions thing causing it.

Anyway, I took another S&R mission, and ran through the procedure.  However, this time, it all worked as expected.  The only difference though, is that I forgot to sell 10t Pesticides from last night.  So I wonder if keeping some cargo keeps the info somehow?  This led on to something else I'll mention below in a minute.

So I sold the Pesticides, and tested my script.  The data was still correct.  I quit EDDI, and reloaded.  Still worked.  I quit Elite, and boom, failure!  So, from this, it would appear that the criteria is: jumping to another system, having no cargo, and quitting Elite.

I'm going to complete this mission, and then test it all again with another one, this time without any cargo when I test, to make sure my hypothesis is correct.  I'll let you know how I get on.

So, back to that other thing I noticed.  Those Pesticides I still had with me.  When I went to sell them, EDDI told me I could sell them for a profit, when the game was telling me I'd make a loss.  A quick bit of further investigation led me to find that EDDI had 'forgotten' the cargo price, and thought it was zero.  This would appear to be related to the above problem of missing cargo data after quitting EDDI and Elite.  I'll see if I can run a few more tests later today to check.

My guess is that somewhere along the line, EDDI stopped writing/updating the cargomonitor.json file, and that is the root cause of all the above.  At least that should give you something to check. :)

EDIT: Corrected 'cargo sell price' to 'cargo price'.

**Comment by Darkcyde13** [comment]
Oh, one other thing... When I was testing selling the cargo to see why I was told I'd make a profit, the FOR loop in 'Commodity sale check' was running through three times; once for the Pesticides, once for the mission cargo (even though I had none), and once more that had blank data.  Shouldn't that loop have only processed once for the cargo I had?  It seems odd that it also did the (non-existent) mission cargo, and even more strange that it did a phantom cargo item too.

**Comment by Darkcyde13** [comment]
So, I've run another mission and made a note for each stage:

> Load EDDI
> Load ED
> Still had some cargo - no data at all, script output non-existent.
> Pickup mission - works (for mission cargo only, the cargo I have was not listed)
> Undock - works (for mission cargo only, the cargo I have was not listed)
> Redock and sold cargo - works
> Jump to system - works
> Jump to 2nd system - works
> Enter normal space at nav beacon - works
> Enter supercruise - works
> Enter normal space - works
> Land at station - works
> Exit & reload EDDI - works
> Exit & reload ED - works
> Exit both & reload EDDI - failed

I'm guessing that its just a happy coincidence that it still works after closing either individually, as it does seem to require both to be closed at the same time.

I'm going to finish this mission as the data is corrupted now, then I'm going to try one more time, and see if quitting both EDDI & ED while still docked at the origin station works or fails.  I'll then try the same at each stage, but this will probably take lot longer with all the reloading.

**Comment by Darkcyde13** [comment]
Right, it appears that my thought about it being related to jumping may be wrong after all.  I just tried another mission and failed at the first hurdle:

> Pick up mission - works for mission only, not the cargo I have
> Undocked, remembered I had cargo, and redocked
> Sold cargo
> Exited both & reload EDDI - failed

But this doesn't explain how on earth it managed to work last night, after rebooting my PC.  I'm at a loss as to why that worked now.

Oh well, I hope that some of this is helpful in determining the cause.  My guess would be the fact that the cargomonitor isn't being written and used anymore.

**Comment by Tkael** [comment]
Strange that cargomonitor.json isn't being written for you. If other players aren't observing the same thing then you might want to reinstall to make sure that you're running the latest version of the .dll file (the file should have approximately the same `Date modified` as EDDI.exe)? 

Assuming that we're not looking at an obsolete .dll file, I can confirm that the file is definitely is being written on my private development branch (which does make some modifications to that module). So perhaps already partially fixed for the next release? 

Can you provide the precise wording from the triple `Commodity sale check`? The wording might provide some clues about the triggering state.

**Comment by Darkcyde13** [comment]
I've just checked my install folder and all files are dated the same, 20th August 2021, so I do have the latest version by the looks of it.

I'm currently about 3000LY out from the bubble, but making my way back tonight.  I'll try a full reinstall (after making a backup) and do some more tests once I'm back.

The thing is, I didn't notice anything wrong until now, a year after the monitor was last modified, and I've done loads of cargo runs and cargo missions, without seeing any problems.  So it's possible that others are also having the problem but not noticed either.  It was only because I've been trying to update my missions scripts, with regards to mission cargo, that I found it to be a problem for me.  99% of the time, I do all my missions/cargo runs in the same game session, and it seems that the cargo data is fine in memory, it has only been when I logged out that it causes any problems.

I'll try to make it back to the bubble tonight if I can, but otherwise it may be a couple of days before I get to try anything out.

**Comment by Darkcyde13** [comment]
So, today I have uninstalled EDDI, and removed the Appdata/Roaming/EDDI folder to remove all saved data.  I then rebooted my PC and reinstalled v4.0.1, and ran it as standalone.  However, the `cargomonitor.json` wasn't created.  I closed EDDI standalone, and ran it as part of VA, just in case that would do anything, but no, still no cargo monitor file.  This is the state of the EDDI folder:

![image](https://user-images.githubusercontent.com/24360725/140648032-946d2f58-3521-480a-84fb-71159a652340.png)

Next, I ran Elite and bought 1t of Hydrogen Fuel, but still no cargo monitor file.  So it looks like there is a problem with the released version of EDDI, and may have been for about a year?  I'm not sure why your dev version is working but the release version is not (at least, not on my PC anyway).  I did one more uninstall/reinstall (without deleting the appdata folder), and was very surprised that the 1t of Hydrogen Fuel was still listed in the Cargo Monitor Tab.  Is this data stored anywhere else, or in a different file?  I checked my whole drive C: for any instances of 'cargomonitor' and only found the Windows references to my search and opening the old file back on the 28th October.  How is EDDI able to remember my cargo if no cargomonitor file is written?  This may be why no one else has noticed it before.

As for the 'triple sales check'.  I added `{dump(cargo.name)}` to the FOR loop (on line 30), and ran the script.  This resulted in: "Hydrogen Fuel" "Black Box" (the mission was a black box salvage mission), but no phantom third entry this time.  After the above testing, it no longer reported the black box mission cargo, and a quick check of my original test code (as in my initial post) showed that the 'need' variable (and others) was again void.  However, after collecting the black boxes and returning to the station (and before hand in) the 'need' variable was again populated with the correct amount.

With a bit more closing/opening of EDDI and Elite, the 'need' variable remained populated.  So I handed in the mission and picked up another, a source & return one this time.  Running my test script resulted in this:
> Mission: Source and return 35 units of CMM Composite. name: CMM Composite. category: Industrial Materials. owned: 0, 0, number. stolen: 0, 0, number. haulage: 0, 0, number. total: 0, 0, number. need: 35, 35, number. price: 0, 0, number.

I tried closing and reopening EDDI and Elite again, and the test remained OK.  'Commodity sale check' (with the `dump()` in) resulted in: "Algae" "Bio-waste" "CMM Composite".  I undocked, redocked, and tested again.  Still OK.  Then I sold the Algae and Biowaste (collected on the previous mission), tested both my test script and 'Commodity sale check', still OK.  So I exited both EDDI and Elite, reloaded EDDI standalone, and it failed again.  My test script gave this:
> Mission: Source and return 35 units of CMM Composite. name:. category:. owned:, , void. stolen:, , void. haulage:, , void. total:, , void. need:, , void. price:, , void.

And 'Commodity sale check' was totally blank, with the Cargo Monitor Tab now completely empty as well.

I am at a loss as to why this seems to happen intermittently, although does seem to happen most of the time when exiting both EDDI and Elite.  I'm also confused as to how EDDI is able to remember the Cargo Monitor data without a cargomonitor.json file, even after closing/opening both, but then will forget it all soon after.

I have tried to install old versions of EDDI, going back to 3.7.0, 3.7.1 and 3.7.2, to see what files get created.  However, in each case EDDI only displayed four tabs: 'EDDI', 'Commander Details', 'Frontier API', and 'Text-to-Speech'.  In the appdata folder there were only four files created: 'eddi.json', 'eddi.log', 'EDDI.sqlite', and 'CompanionAPI.json', so I am unable to check if cargomonitor would be created.  I seem to remember this 'four tabs' thing was a known issue some time ago, so I guess I'm just suffering that problem with this.
